### PR TITLE
move output file name logic into core

### PR DESCRIPTION
**Pull request by beauxq**
I see the same logic with small variations in each different world implementation. It seems to me, it would be better in the core to keep it consistent.

This change is not applied to DS3, because there's a different issue there that can be handled in a different pull request.
(DS3 relies on the unsafe-for-filename player name being in the filename, and is parsed in [WebHostLib/upload.py](https://github.com/ArchipelagoMW/Archipelago/blob/36eb80c725f844fa62101a40795ed7570494d281/WebHostLib/upload.py#L83-L84))

## What is this fixing or adding?

move output file name logic from most games into core

## How was this tested?

 - unit tests
 - generate a 2-player super metroid seed

Maybe a few people could test generate for other rom games.

### CODE REVIEW

File: BaseClasses.py [hunk 1]
```diff
@@ -294,9 +294,18 @@ def get_name_string_for_object(self, obj) -> str:
     def get_player_name(self, player: int) -> str:
         return self.player_name[player]
 
+    # REVIEW: Change this to `_get_file_safe_player_name` ?
```
Comment (Berserker66) [inline BaseClasses.py:297]: can probably remain a public API in case someone has a use for it.

File: worlds/alttp/__init__.py [hunk 2]
```diff
@@ -396,11 +396,7 @@ def generate_output(self, output_directory: str):
                                deathlink=world.death_link[player],
                                allowcollect=world.allow_collect[player])
 
-            outfilepname = f'_P{player}'
-            outfilepname += f"_{world.get_file_safe_player_name(player).replace(' ', '_')}" \
-                if world.player_name[player] != 'Player%d' % player else ''
-
-            rompath = os.path.join(output_directory, f'AP_{world.seed_name}{outfilepname}.sfc')
+            rompath = os.path.join(output_directory, f'{self.world.get_out_file_name_base(self.player)}.sfc')
```
Comment (black-sliver) [inline worlds/alttp/__init__.py:399]:
```suggestion
            rompath = os.path.join(output_directory, f"{self.world.get_out_file_name_base(self.player)}.sfc")
```

**Review by black-sliver — CHANGES_REQUESTED** [review]
Suggesting quotes as per https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/style.md#python-code where touched

**Comment by black-sliver** [comment]
The json name change crashes here in webhost:
https://github.com/ArchipelagoMW/Archipelago/blob/36eb80c725f844fa62101a40795ed7570494d281/WebHostLib/upload.py#L83-L84

Pinging @Marechal-L for naming suggestions

**Comment by beauxq** [comment]
> The json name change crashes here in webhost

I mentioned small variations. One of the variations I saw was hyphen `-` instead of underscore `_`.

I wonder if this just needs `split('-'` to change to `split('_'` ?

**Comment by beauxq** [comment]
It looks like DS3 was not using the filename-safe version of the player name, which means it might have crashed (under the current system) if someone used certain characters in their name?

If we need the actual player name, rather than the one that has been made safe for file names, I think that should probably come from inside the json data, rather than from the filename.

**Comment by black-sliver** [comment]
> I wonder if this just needs split('-' to change to split('_' ?

This will break the transition from 0.3.5 to 0.3.6, which is *probably* not that much of a deal (besides not getting a good error message), but I don't know if that change would require changes elsewhere (i.e. in the game documentation), which is why i pinged the world maintainer.

>  It looks like DS3 was not using the filename-safe version of the player name, which means it might have crashed (under the current system) if someone used certain characters in their name?
> And what if someone puts a hyphen in their name?

split(..., 3) stops splitting at the player_name, so that's no issue. Unsafe player_name however is

> If we need the actual player name, rather than the one that has been made safe for file names, I think that should probably come from inside the json data, rather than from the filename.

Yeah, i think it should first read multidata and set up the slots, then iterate over stuff and set up the downloads (validating game, not setting game).

As a side note, we are not super happy with using *.json. The preferred thing would be to provide a APContainer for all games where feasible, but adding unzip to all game clients is not easy to do. I think APContainer support could be part of apclientpp, but I want to keep that header-only to avoid the dependency mess.

**Comment by Marechal-L** [comment]
I worked on a feature to remove the output file and use the slot_data instead. I will open the pull-request this weekend.
So feel free to fix it as simple as possible

**Comment by beauxq** [comment]
I reverted the change to DS3 output file name, so that issue can be handled in a different pull request, and this doesn't break it more.

**Review by black-sliver — APPROVED** [review]

**Pull request merged**